runtime/memory.c

```diff
@@ -488,25 +488,17 @@ CAMLexport value caml_alloc_shr_noexc(mlsize_t wosize, tag_t tag) {
    the implementation from the user.
 */
 
-/* A type with the most strict alignment requirements */
-union max_align {
-  char c;
-  short s;
-  long l;
-  int i;
-  float f;
-  double d;
-  void *v;
-  void (*q)(void);
-};
+#if !defined(HAVE_MAX_ALIGN_T) && defined(_MSC_VER)
+typedef double max_align_t;
+#endif
 
 struct pool_block {
 #ifdef DEBUG
   intnat magic;
 #endif
   struct pool_block *next;
   struct pool_block *prev;
-  union max_align data[];  /* not allocated, used for alignment purposes */
+  max_align_t data[]; /* not allocated, used for alignment purposes */
 };
```

Comment:
I'd tend to align (ha, ha) with your suggestion, but: MSVC C++ `cstddef.h` header uses double alignment for `max_align_t`, and clang-cl uses double too, which I think makes it a reasonable default for Windows.
As for a general fallback, I hope that other compilers+libc aren't as buggy, and I wonder if we do need to provide a definition, or rather catch a missing definition as a compilation failure.

Reply:
> MSVC C++ cstddef.h header uses double alignment for max_align_t

It is in error, then.  The program below, compiled with MSVC, prints 16, showing that there are types with alignment greater than 8.
```
#include <iostream>
#include <xmmintrin.h>

int main() {
    std::cout << alignof(__m128) << '\n';
    return 0;
}
```
In this PR, you're not trying to emulate whatever strange choices MSVC does, but to make sure OCaml's pool allocator works as intended.  If the intent is to align for the maximal alignment constraint of the target platform, the alignment must be >= 16 on x86, because SSE instructions.  If the intent is to align for the biggest datatype OCaml stores in heap blocks, word-alignment is enough and you don't need to add anything to `struct pool_block` to guarantee it, as it already contains two word-sized pointers.

Reply:
What about defaulting to `long double` rather than `double` then?

Reply:
> It is in error, then. The program below, compiled with MSVC, prints 16, showing that there are types with alignment greater than 8.

At least for C++, I don't think there's an error.

> The type `max_align_t` is a POD type whose alignment requirement is at least as great as that of every scalar type, and whose alignment requirement is supported in every context. [C++11]

```c++
#include <iostream>
#include <xmmintrin.h>
#include <cstddef>
#include <type_traits>

int main() {
    std::cout << alignof(__m128) << std::endl
              << alignof(std::max_align_t) << std::endl
              << std::is_scalar<__m128>() << std::endl;
    return 0;
}
```
shows `16`, `8`, `0`, indicating that `__m128` isn't considered a scalar type, and thus the definition of `std::max_align_t` is consistent (with respect to the `__m128` type).

The definition for C is more vague and I guess it could be argued that `16` would be a correct value.

> `max_align_t` which is an object type whose alignment is as great as is supported by the implementation in all contexts; [C11]

> What about defaulting to long double rather than double then?

`long double` is identical to `double` under MSVC [^1].

Thanks for the thorough review. I think the real question is indeed whether the intent is to align for the maximal alignment constraint of the target platform, or to align for the biggest datatype OCaml stores in heap blocks. None of the fields in the former `union max_align` had a 16 bytes alignment, and all worked well, hasn't it? Now it's not clear to me that this field is actually needed.

[^1]: https://learn.microsoft.com/en-us/cpp/c-language/type-long-double?view=msvc-170

Reply:
See also #12212.  My guess is that allocation pools are realigned no matter how `struct pool_block` is defined.

Reply:
"this comment" refers to the OCaml 4 memory allocator, which has been extensively rewritten for OCaml 5.  Please see #12212 and the corresponding OCaml 5 code to check how alignment is actually handled in pool blocks.

Reply:
I have the impression that you are talking about two different things, which very conveniently have the exact same name in the codebase. In memory.c (which the PR is hacking on), "pool" refers to a single global pool of blocs (`struct pool_block`) that have been `caml_stat_alloc`ed by the runtime, and is used to make sure that they are all freed on program termination -- I think that it is only used in cleanup-at-exit mode? There is a single memory pool, which is basically a doubly-linked list, and each `caml_stat_alloc` adds one element to it.

In shared_heap.c, `struct pool` refers to one block of memory in the memory, of size 4096 words, and owned by a domain-local `caml_heap_state` structure. I think that the alignment constraints that @xavierleroy has in mind apply to values stored in the shared_heap.c pools, but that those are in fact currently not stored in the global memory.c pool, as they are allocated by `caml_mem_map` in shared_heap.c:pool_acquire. (Large objects are allocated differently in `large_allocate`, and oddly enough they seem to use `malloc` and not `caml_stat_alloc`.)

Reply:
Thanks for the clarification, it makes more sense now.  

So, we're talking about `caml_stat_alloc`, which should have the same good properties as `malloc`, in particular: align sufficiently so that all data types supported by the target architecture (not just those used by the OCaml runtime system, because third-party OCaml-C stubs) can be stored safely.  For x86, it means 16-alignment, because of SSE 128-bit vector operations.  (256- and 512-bit AVX operations support less aligned accesses, albeit slowly.)  For the other OCaml target architectures, 8-alignment seems enough, e.g. for ARM64's load/store register pair instructions, and for ARM NEON, but I'm not sure.

(Note that I wrote "stored safely", i.e. the program doesn't crash, not "stored efficiently" with the alignment that gives best hardware performance for vector instructions.  The latter may require specific allocators besides `malloc`, for the reason below.)

Giving `struct pool_block` an alignment greater than the one guaranteed by `malloc` is useless.  E.g. if `struct pool_block` is 16-aligned and `malloc` returns an address = 8 mod 16 (as can happen in 32-bit Windows, I heard), you'll add 16 to this address and get something that is not 16 aligned, but 8 bytes are wasted.

For this reason, I don't believe in the "align on cache lines" argument of "this comment": cache lines are typically 64-byte wide, sometimes 128 or even 256.  `malloc` will not return 64-byte aligned pointers, at least not for small- to medium-sized blocks, because it would waste too much space, and trying to realign to 64 afterward would waste much space too.  

What does this mean for this PR?  
- The target systems we care about are 64-bit architecture with alignment constraints <= 16 and malloc returning 16-aligned blocks.  In this case, the `data` part of `struct pool_block` is naturally 16-aligned (because the two pointers before use 16 bytes), and nothing needs to be done.  Aligning `data` using `max_align_t` should have no effect. (*)
- For a 32-bit architecture with 16-byte alignment constraints and malloc returning 16-aligned blocks (e.g. Linux x86-32), aligning `data` to 16 seems preferable to me and can be achieved by using `max_align_t`.
- For a 32-bit architecture with 16-byte alignment constraints and malloc returning 8-aligned blocks (perhaps Windows 32 bits, not sure): no amount of alignment constraints in `struct pool_block` will give 16-aligned `data` fields, so you could just as well put no alignment constraints.

(*) There's still an issue with the pesky magic number added in debug mode, which throws alignment off.  I wish we would just remove it, as I don't think it adds anything to the debug mechanisms built into most malloc implementations.

That's all (and everything) I had to say about this PR.  Now I'm done with this discussion and this PR.

Reply:
Thank you for the review and the explanations, there's a lot to be learned here.